File: 2023/day15/solution.py
#!/usr/bin/env python3
import sys
from functools import cache
import re
from collections import defaultdict

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case '-p1':
        PART1, PART2 = True, False
      case '-p2':
        PART1, PART2 = False, True
      case '-p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  inputs = ''.join(f.read().splitlines()).split(',')


def hashfn(val: str) -> int:
  res = 0
  for x in val:
    res += ord(x)
    res *= 17
    res %= 256
  return res


def part1():
  total = 0
  for x in inputs:
    total += hashfn(x)
  print(f'part 1: {total}')


def part2():
  # did a high person write this question
  regex = re.compile(r'([a-z]+)([-=])([0-9]?)')
  # python dict preserves insertion order. solves in other languages will be fun
  boxes: defaultdict[int, dict[str, int]] = defaultdict(lambda: dict())
  for x in inputs:
    match = re.match(regex, x)
    assert match
    label = match.group(1)
    op = match.group(2)
    lens = int(match.group(3)) if match.group(3) else None
    box = hashfn(label)
    if DEBUG > 1: print(f'{x=} {label=} {op=} {lens=} {box=}')
    match op:
      case '=':
        assert lens
        boxes[box][label] = lens
      case '-':
        # delete lens. delete box if empty
        if label in boxes.get(box, {}):
          del boxes[box][label]
          if not boxes[box]:
            del boxes[box]
      case _:
        assert False, f'{op=}'
    if DEBUG > 1: print(f'  {dict(boxes)}')

  total = 0
  for box_id, lenses in boxes.items():
    for slot_id, lens in enumerate(lenses.values()):
      value = (box_id + 1) * (slot_id + 1) * lens
      if DEBUG > 0: print(f'{box_id=} {slot_id=} {lens=} {value=}')
      total += value
  print(f'part 2: {total}')


if PART1: part1()
if PART2: part2()

File: 2023/day15/solution.ts
#!/usr/bin/env -S deno --allow-read
// #region base aoc template
class Args {
  constructor(public filename = 'input.txt', public debug = 0, public part1 = true, public part2 = true) {
    for (const arg of Deno.args) {
      const [key, value] = [arg.slice(0, 2), (fallback = 0) => Number(arg.slice(2) || fallback)];
      if (key === '-d') debug = value(1);
      else if (key === '-e') filename = `example${arg.slice(2)}.txt`;
      else if (key === '-i') filename = 'input.txt';
      else if (key === '-p') [part1, part2] = [(value(0) & 1) === 1, (value(0) & 2) === 2];
      else throw new Error(`unrecognised arg="${arg}"`);
    }
    [this.filename, this.debug, this.part1, this.part2] = [filename, debug, part1, part2];
    console.log(`args: {filename: "${filename}", debug: ${debug}, part1: ${part1}, part2: ${part2} }`);
  }
}

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (args.debug >= level) console.debug(...data);
};

const args = new Args();
const input = await Deno.readTextFile(args.filename);
// #endregion

const parseInput = () =>
  input
    .replaceAll(/\s+/g, '')
    .split(',')
    .filter((token) => token.trim());

const hash = (data: string) =>
  data.split('').reduce((acc, char) => {
    acc += char.charCodeAt(0);
    acc *= 17;
    acc &= 255;
    return acc;
  }, 0);

const part1 = () =>
  console.log(
    'part 1:',
    parseInput().reduce((acc, item) => acc + hash(item), 0)
  );

const part2 = () => {
  interface lens {
    label: string;
    focal: number;
  }
  const boxes = new Map<number, lens[]>();
  const regex = /([a-z]+)([-=])([0-9])?/;
  for (const code of parseInput()) {
    const [_, label, operation, focalStr] = regex.exec(code)!;
    const box = hash(label);
    debug(2, { label, operation, focalStr, box });
    if (operation === '=') {
      const focal = Number(focalStr);
      if (!boxes.has(box)) {
        // create
        boxes.set(box, [{ label, focal }]);
      } else if ((boxes.get(box) ?? []).filter((item) => item.label === label).length > 0) {
        // replace
        boxes.set(
          box,
          boxes.get(box)!.map((item) => (item.label === label ? { label, focal } : item))
        );
      } else {
        // push back
        boxes.get(box)!.push({ label, focal });
      }
    } else if (operation === '-') {
      // why wont ?. work :@
      if ((boxes.get(box) ?? []).filter((item) => item.label === label).length > 0)
        boxes.set(
          box,
          boxes.get(box)!.filter((item) => item.label !== label)
        );
    }
    debug(3, { boxes });
  }
  debug(1, { boxes });
  const power = boxes
    .entries()
    .reduce((acc, [box, lenses]) => acc + lenses.reduce((acc, lens, i) => acc + (box + 1) * (i + 1) * lens.focal, 0), 0);
  console.log('part 2:', power);
};

if (args.part1) part1();
if (args.part2) part2();
